### DISCUSSION TITLE
Ending a controller with a get_p call

### DISCUSSION THREAD

**Discussion by daleif** [mojolicious/mojo#2099] [Q&A]
This has been annoying me for a while now, and it bugs me that it is listed in the FAQ, but with no explanation on how to solve it.

Say I have controller that is suppose to fetch a page online, get some info and return some JSON. Why does this not work (I replaced the return statements with

```
sub extract ($self) {
  my $url   = 'SOM EURL';
  my $ua    = Mojo::UserAgent->new->inactivity_timeout(15)->max_redirects(5);

  $self->render_later;
  
  return $ua->get_p($url)
      ->then(sub ($tx)  {
		 $self->log->info($self->dumper($tx));
                 return $self->render(json => {title => 'test'});
	     })
      ->catch( sub ($err) {
		   $self->log->error($err);
                  return $self->render(json => undef);
	       })->wait;
}
```

I keep getting "Premature connection close". I get that something runs out of scope. But when what exactly is the paradigm we should use for something like this?

**Comment by kraih** [comment]
You need to prevent `$ua` from going out of scope. Store it in a helper or so.

**Reply by daleif** [reply]
It would be nice if there were an example of a solution do such a thing somewhere in the documentation.

I added an our variable in my plugin where i store my helpers and made a helper that just sets that variable to $ua. That works. Using a my variable did not.

**Reply by Grinnz** [reply]
https://docs.mojolicious.org/Mojolicious/Guides/FAQ#What-does-Premature-connection-close-mean

**Reply by Grinnz** [reply]
Alternatively to an our variable, you can store the $ua in a helper or at a higher scope in the file.

**Reply by daleif** [reply]
I know I've read it, but it has now good suggestions for solutions.

I did try storing it in a helper, but that gave the same result, hence the our.

**Comment by hussam-qasem** [comment]
Perhaps something like this?

```perl
use Mojolicious::Lite -signatures, -async_await;

helper my_ua => sub {
  use Mojo::UserAgent;
  my $ua = Mojo::UserAgent->new->inactivity_timeout(15)->max_redirects(5);
  $ua
};

any '/' => async sub ($c) {
    my $url = Mojo::URL->new('https://httpbin.org/json');
    $c->render_later();
    my $results = await Mojo::Promise->all_settled( $c->my_ua->get_p( $url ) );
    if ($results->{status} eq "fulfilled") {
      my $tx = $results->{value}[0];
      my $json = $tx->res->json;
      $c->render(json => $json);
    }
    else {
      my $reason = $results->{reason}[0];
      $c->render(json => {reason => $reason});
    }
};

app->start;
```

Also, if you're proxying requests consider the [`proxy`](https://docs.mojolicious.org/Mojolicious/Plugin/DefaultHelpers#proxy-get_p) helper.

**Reply by daleif** [reply]
Very interesting and a better helper than what I was doing ;-) Will do some tests.

**Reply by hussam-qasem** [reply]
You can also use the built-in user agent, e.g. `$c->ua->get_p( $url )` and control parameter using [environment variables](https://docs.mojolicious.org/Mojo/UserAgent#inactivity_timeout).

**Reply by daleif** [reply]
I thought I read somewhere that one should not mess with `$c->ua`. 
BTW: do you have any ideas for https://github.com/mojolicious/mojo/discussions/2090

**Reply by hussam-qasem** [reply]
It's there in the [docs](https://docs.mojolicious.org/Mojolicious/Guides/Tutorial#User-agent)
